File: backend/routes/admin/vouchers.ts
import { Router, Request, Response } from 'express';
import db from '../../config/database';

const router = Router();

// GET /api/admin/vouchers
router.get('/', async (_req: Request, res: Response) => {
  try {
    const result = await db.query('SELECT * FROM vouchers ORDER BY created_at DESC');
    return res.json({ success: true, vouchers: result.rows });
  } catch (err) {
    return res.status(500).json({ success: false, error: (err as Error).message });
  }
});

// POST /api/admin/vouchers
router.post('/', async (req: Request, res: Response) => {
  const { code, discount_type, discount_value, max_uses, expires_at, is_active } = req.body;
  if (!code || !discount_type || discount_value == null) {
    return res.status(400).json({ success: false, error: 'code, discount_type and discount_value are required' });
  }
  try {
    const result = await db.query(
      `INSERT INTO vouchers (code, discount_type, discount_value, max_uses, expires_at, is_active)
       VALUES ($1, $2, $3, $4, $5, $6)
       RETURNING *`,
      [code.toUpperCase(), discount_type, discount_value, max_uses ?? null, expires_at ?? null, is_active ?? true]
    );
    return res.json({ success: true, voucher: result.rows[0] });
  } catch (err) {
    return res.status(500).json({ success: false, error: (err as Error).message });
  }
});

// PUT /api/admin/vouchers/:id
router.put('/:id', async (req: Request, res: Response) => {
  const { id } = req.params;
  const { discount_type, discount_value, max_uses, expires_at, is_active } = req.body;
  try {
    const result = await db.query(
      `UPDATE vouchers
       SET discount_type = $1, discount_value = $2, max_uses = $3, expires_at = $4, is_active = $5
       WHERE id = $6
       RETURNING *`,
      [discount_type, discount_value, max_uses ?? null, expires_at ?? null, is_active, id]
    );
    if (result.rows.length === 0) {
      return res.status(404).json({ success: false, error: 'Voucher not found' });
    }
    return res.json({ success: true, voucher: result.rows[0] });
  } catch (err) {
    return res.status(500).json({ success: false, error: (err as Error).message });
  }
});

// DELETE /api/admin/vouchers/:id
router.delete('/:id', async (req: Request, res: Response) => {
  const { id } = req.params;
  try {
    await db.query('DELETE FROM vouchers WHERE id = $1', [id]);
    return res.json({ success: true });
  } catch (err) {
    return res.status(500).json({ success: false, error: (err as Error).message });
  }
});

export default router;

File: backend-golang/internal/handlers/admin/vouchers.go
package admin

import (
	"context"
	"encoding/json"
	"net/http"
	"strings"

	"github.com/go-chi/chi/v5"

	"inanhxink/backend-golang/internal/config"
	"inanhxink/backend-golang/internal/handlers"
)

// GET /api/admin/vouchers
func ListVouchers(w http.ResponseWriter, r *http.Request) {
	rows, err := config.DB.Query(context.Background(),
		"SELECT * FROM vouchers ORDER BY created_at DESC")
	if err != nil {
		handlers.InternalError(w, err)
		return
	}
	vouchers, err := handlers.CollectRows(rows)
	if err != nil {
		handlers.InternalError(w, err)
		return
	}
	handlers.OK(w, map[string]any{"success": true, "vouchers": vouchers})
}

// POST /api/admin/vouchers
func CreateVoucher(w http.ResponseWriter, r *http.Request) {
	var body map[string]any
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		handlers.BadRequest(w, "Invalid JSON")
		return
	}
	code, _ := body["code"].(string)
	discountType, _ := body["discount_type"].(string)
	discountValue := body["discount_value"]
	if code == "" || discountType == "" || discountValue == nil {
		handlers.BadRequest(w, "code, discount_type and discount_value are required")
		return
	}
	isActive := true
	if v, ok := body["is_active"].(bool); ok {
		isActive = v
	}

	rows, err := config.DB.Query(context.Background(), `
		INSERT INTO vouchers (code, discount_type, discount_value, max_uses, expires_at, is_active)
		VALUES ($1,$2,$3,$4,$5,$6) RETURNING *`,
		strings.ToUpper(code), discountType, discountValue, body["max_uses"], body["expires_at"], isActive)
	if err != nil {
		handlers.InternalError(w, err)
		return
	}
	row, _ := handlers.CollectOne(rows)
	handlers.OK(w, map[string]any{"success": true, "voucher": row})
}

// PUT /api/admin/vouchers/:id
func UpdateVoucher(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")
	var body map[string]any
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		handlers.BadRequest(w, "Invalid JSON")
		return
	}
	rows, err := config.DB.Query(context.Background(), `
		UPDATE vouchers SET discount_type = $1, discount_value = $2, max_uses = $3,
			expires_at = $4, is_active = $5
		WHERE id = $6 RETURNING *`,
		body["discount_type"], body["discount_value"], body["max_uses"],
		body["expires_at"], body["is_active"], id)
	if err != nil {
		handlers.InternalError(w, err)
		return
	}
	row, err := handlers.CollectOne(rows)
	if err != nil || row == nil {
		handlers.NotFound(w)
		return
	}
	handlers.OK(w, map[string]any{"success": true, "voucher": row})
}

// DELETE /api/admin/vouchers/:id
func DeleteVoucher(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")
	if _, err := config.DB.Exec(context.Background(),
		"DELETE FROM vouchers WHERE id = $1", id); err != nil {
		handlers.InternalError(w, err)
		return
	}
	handlers.OK(w, map[string]any{"success": true})
}
